Firestore/core/src/local/leveldb_target_cache.cc

```diff
@@ -188,35 +189,56 @@ absl::optional<TargetData> LevelDbTargetCache::GetTarget(const Target& target) {
   return absl::nullopt;
 }
 
-void LevelDbTargetCache::EnumerateTargets(const TargetCallback& callback) {
+void LevelDbTargetCache::EnumerateSequenceNumbers(
+    const SequenceNumberCallback& callback) {
   // Enumerate all targets, give their sequence numbers.
   std::string target_prefix = LevelDbTargetKey::KeyPrefix();
   auto it = db_->current_transaction()->NewIterator();
   it->Seek(target_prefix);
   for (; it->Valid() && absl::StartsWith(it->key(), target_prefix);
        it->Next()) {
-    TargetData target = DecodeTarget(it->value());
-    callback(target);
+    StringReader reader{it->value()};
+    auto target_proto = DecodeTargetProto(&reader);
+    callback(target_proto->last_listen_sequence_number);
   }
 }
 
-int LevelDbTargetCache::RemoveTargets(
+uint64_t LevelDbTargetCache::RemoveTargets(
     ListenSequenceNumber upper_bound,
     const std::unordered_map<model::TargetId, TargetData>& live_targets) {
-  int count = 0;
   std::string target_prefix = LevelDbTargetKey::KeyPrefix();
   auto it = db_->current_transaction()->NewIterator();
   it->Seek(target_prefix);
+
+  std::unordered_set<model::TargetId> removed_targets;
+
+  // In https://github.com/firebase/firebase-ios-sdk/issues/6721, a customer
+  // reports that their client crashes when deserializing an invalid Target
+  // during an LRU run. Instead of deserializing the value into a full Target
+  // model, we only convert it into the underlying Protobuf message.
   for (; it->Valid() && absl::StartsWith(it->key(), target_prefix);
        it->Next()) {
-    TargetData target_data = DecodeTarget(it->value());
-    if (target_data.sequence_number() <= upper_bound &&
-        live_targets.find(target_data.target_id()) == live_targets.end()) {
-      RemoveTarget(target_data);
-      count++;
+    StringReader reader{it->value()};
+    auto target_proto = DecodeTargetProto(&reader);
+    if (target_proto->last_listen_sequence_number <= upper_bound &&
+        live_targets.find(target_proto->target_id) == live_targets.end()) {
+      TargetId target_id = target_proto->target_id;
+
+      RemoveAllDocumentKeysForTarget(target_id);
+
+      std::string key = LevelDbTargetKey::Key(target_id);
```

Comment:
This is the removal of the key from the LevelDb `LevelDbTargetKey` map (Target Id to Target Proto). this maps to the following in RemoveTarget:

```
  std::string key = LevelDbTargetKey::Key(target_id);
  db_->current_transaction()->Delete(key);
```

`RemoveQueryTargetKeyForTargets` does what corresponds to your code snippet.  It iterates all "canonical ID to target ID" mappings and removes the ones that match the target ID. This is done because we don't know the canonical ID.

I added some comments.